block/bench/Block.hs

```diff
@@ -0,0 +1,249 @@
+    , bvdUpdateImplicit = 0
+    , bvdSoftforkRule     = SoftforkRule
+          { srInitThd      = unsafeCoinPortionFromDouble 0
+          , srMinThd       = unsafeCoinPortionFromDouble 0
+          , srThdDecrement = unsafeCoinPortionFromDouble 0
+          }
+    , bvdTxFeePolicy      = TxFeePolicyUnknown 0 mempty
+    , bvdUnlockStakeEpoch = EpochIndex { getEpochIndex = 0 }
+    }
+  where
+    limit = fromIntegral ((2 :: Int) ^ (32 :: Int))
+
+genesisInitializer :: GenesisInitializer
+genesisInitializer = GenesisInitializer
+    { giTestBalance = balance
+    , giFakeAvvmBalance = FakeAvvmOptions
+          { faoCount = 1
+          , faoOneBalance = maxBound
+          }
+    , giAvvmBalanceFactor = unsafeCoinPortionFromDouble 0
+    , giUseHeavyDlg = False
+    , giSeed = 0
+    }
+
+balance :: TestnetBalanceOptions
+balance = TestnetBalanceOptions
+    { tboPoors = 1
+    , tboRichmen = 1
+    , tboTotalBalance = maxBound
+    , tboRichmenShare = 1
+    , tboUseHDAddresses = False
+    }
+
+genesisSpec :: GenesisSpec
+genesisSpec = UnsafeGenesisSpec
+    { gsAvvmDistr = GenesisAvvmBalances mempty
+    , gsFtsSeed = SharedSeed mempty
+    , gsHeavyDelegation = UnsafeGenesisDelegation mempty
+    , gsBlockVersionData = bvd
+    , gsProtocolConstants = pc
+    , gsInitializer = genesisInitializer
+    }
+
+confDir :: FilePath
+confDir = "./lib"
+
+-- | A test subject: a MainBlock, and its various components, each paired with
+-- its serialization.
+data TestSubject = TestSubject
+    { tsBlock       :: !(MainBlock, LBS.ByteString)
+    , tsHeader      :: !(MainBlockHeader, LBS.ByteString)
+    , tsBodyProof   :: !(BodyProof MainBlockchain, LBS.ByteString)
+    , tsConsensus   :: !(ConsensusData MainBlockchain, LBS.ByteString)
+    , tsExtraHeader :: !(ExtraHeaderData MainBlockchain, LBS.ByteString)
+    , tsBody        :: !(Body MainBlockchain, LBS.ByteString)
+    , tsTxPayload   :: !(TxPayload, LBS.ByteString)
+    , tsSscPayload  :: !(SscPayload, LBS.ByteString)
+    , tsDlgPayload  :: !(DlgPayload, LBS.ByteString)
+    , tsUpdPayload  :: !(UpdatePayload, LBS.ByteString)
+    , tsExtraBody   :: !(ExtraBodyData MainBlockchain, LBS.ByteString)
+    }
+
+instance NFData TestSubject where
+    rnf (TestSubject {..}) =
+        deepseq tsBlock $
+          deepseq tsHeader $
+          deepseq tsBodyProof $
+          deepseq tsConsensus $
+          deepseq tsExtraHeader $
+          deepseq tsBody $
+          deepseq tsTxPayload $
+          deepseq tsSscPayload $
+          deepseq tsDlgPayload $
+          deepseq tsUpdPayload $
+          deepseq tsExtraBody $
+          ()
+
+testSubjectSize :: (a, LBS.ByteString) -> Int64
+testSubjectSize (_, bytes) = LBS.length bytes
+
+printSize :: TestSubject -> String -> (TestSubject -> (a, LBS.ByteString)) -> IO ()
+printSize ts name f = Universum.putStrLn $ name <> " of size " <> Universum.show (testSubjectSize (f ts)) <> " bytes"
+
+printSizes :: TestSubject -> IO TestSubject
```

Comment:
Why isn't it `IO ()` then?

Reply:
Because I want to get the subjects back, for ease of use in `Criterion.env`.